File: iomerge-server-app/src/main/java/pl/kbieron/iomerge/server/utils/plugins/PluginLoader.java
package pl.kbieron.iomerge.server.utils.plugins;

import org.pmw.tinylog.Logger;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;
import org.springframework.core.io.support.PropertiesLoaderUtils;
import org.springframework.util.StringUtils;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Properties;
import java.util.function.Function;
import java.util.stream.Stream;

import static pl.kbieron.iomerge.server.config.ConstantPaths.PLUGINS_DIR;


public class PluginLoader {

	private static final String SPRING_CONFIGURATION_PROP = "plugin.spring.configuration";

	private final List<ClassLoader> classLoaders;

	public PluginLoader() {
		classLoaders = new ArrayList<>();
		classLoaders.add(Thread.currentThread().getContextClassLoader());
	}

	public void loadPluginsToContext(AnnotationConfigApplicationContext ctx) {
		if (!PLUGINS_DIR.isDirectory()) {
			Logger.info("Plugin dir doesn't exists: ", PLUGINS_DIR);
			return;
		}

		getPluginsJars()
				.flatMap(this::loadConfigClass)
				.forEach(ctx::register);

		ctx.setClassLoader(new ChainedClassLoader(classLoaders));

	}

	/**
	 * @return stream of jar files in  plugins directory
	 */
	private Stream<File> getPluginsJars() {
		File[] plugins = PLUGINS_DIR.listFiles(file -> file.isFile() && file.getName().endsWith(".jar"));
		return (plugins != null) ? Arrays.stream(plugins) : Stream.empty();

	}


	/**
	 * @return stream of configuration classes in given plugin jar file
	 */
	private Stream<Class<?>> loadConfigClass(File pluginJar) {
		try {
			ClassLoader pluginClassLoader = new URLClassLoader(
					new URL[]{pluginJar.toURI().toURL()},
					Thread.currentThread().getContextClassLoader());

			// load classes
			Stream<Class<?>> classStream = getConfigurationClassNames(pluginClassLoader)
					.map(mapToClassObjects(pluginClassLoader));

			classLoaders.add(pluginClassLoader);

			return classStream;

		} catch (Exception ex) {
			Logger.warn(ex, "Problem while loading plugin");
			return Stream.empty();
		}
	}

	private Stream<String> getConfigurationClassNames(ClassLoader pluginClassLoader) throws IOException {
		Properties properties = PropertiesLoaderUtils.loadAllProperties("iomerge-plugin.properties", pluginClassLoader);
		String classNames = properties.getProperty(SPRING_CONFIGURATION_PROP);

		if (StringUtils.isEmpty(classNames))
			throw new IllegalArgumentException("Plugin mendatory property: '" + SPRING_CONFIGURATION_PROP + "' not found");

		// split class names by ',' and map to class stream
		return Arrays.stream(classNames.split(","));
	}

	private Function<String, Class<?>> mapToClassObjects(ClassLoader pluginClassLoader) {
		return className -> {
			try {
				return pluginClassLoader.loadClass(className);
			} catch (ClassNotFoundException e) {
				throw new RuntimeException(e);
			}
		};
	}

}

File: iomerge-server-app/src/main/java/com/github/krzychek/iomerge/server/utils/plugins/PluginLoader.kt
package com.github.krzychek.iomerge.server.utils.plugins

import com.github.krzychek.iomerge.server.api.PluginProperties.PLUGIN_CLASSES_PROP
import com.github.krzychek.iomerge.server.api.appState.AppStateManager
import com.github.krzychek.iomerge.server.api.network.MessageDispatcher
import com.github.krzychek.iomerge.server.config.AppConfigurator.Paths.pluginsDir
import com.github.krzychek.iomerge.server.network.MessageDispatcherImpl
import org.pmw.tinylog.Logger
import org.springframework.core.io.support.PropertiesLoaderUtils
import org.springframework.stereotype.Component
import java.io.File
import java.net.URLClassLoader
import java.util.*


@Component
open class PluginLoader(appStateManager: AppStateManager,
						messageDispatcher: MessageDispatcherImpl) {

	private val injectableObjects = mapOf(
			AppStateManager::class.java to appStateManager,
			MessageDispatcher::class.java to messageDispatcher
	)

	private val pluginClasses = readPluginJars(pluginsDir)
			.flatMap { loadPluginClasses(it) }

	private val classToObject = WeakHashMap<Class<*>, Any>()

	/**
	 * @return stream of jar files in  plugins directory
	 */
	private fun readPluginJars(dir: File): Array<File> {
		return dir.listFiles { file -> file.isFile && file.name.endsWith(".jar") }
				?: emptyArray<File>()
	}

	/**
	 * create (or gets from cache, if already created) plugin objects of particular supertype
	 */
	fun <T> getPluginObjectsOfType(type: Class<T>): List<T>
			= pluginClasses
			.filter { type.isAssignableFrom(it) }
			.map { @Suppress("UNCHECKED_CAST") (it as Class<out T>) }
			.map { getObjectOfType(it) }


	/**
	 * creates objects of particular class or return if was created before
	 */
	private fun <T> getObjectOfType(clazz: Class<T>): T = clazz.cast(
			classToObject[clazz] ?:
					clazz.constructors.maxBy { it.parameterCount }!!.run {
						val parameters = parameterTypes.map { injectableObjects[it] }.toTypedArray()
						newInstance(*parameters)
					}
	)


	/**
	 * @return stream of configuration classes in given plugin jar file
	 */
	private fun loadPluginClasses(pluginJar: File): List<Class<*>> {
		try {
			val pluginClassLoader = URLClassLoader(
					arrayOf(pluginJar.toURI().toURL()),
					Thread.currentThread().contextClassLoader)

			// load classes
			return pluginClassLoader.getConfigurationClassNames()
					.map { pluginClassLoader.loadClass(it) }

		} catch (ex: Exception) {
			Logger.warn(ex, "Problem while loading plugin ", pluginJar)
			return emptyList()
		}
	}

	private fun ClassLoader.getConfigurationClassNames(): List<String> {
		val properties = PropertiesLoaderUtils.loadAllProperties("iomerge-plugin.properties", this)
		val classNames = properties.getProperty(PLUGIN_CLASSES_PROP)

		if (classNames.isNullOrEmpty())
			throw IllegalArgumentException("Plugin mendatory property: '$PLUGIN_CLASSES_PROP' not found")

		return classNames.split(",").filter { !it.isEmpty() }
	}
}
